src/allocator_defrag.c

```diff
@@ -0,0 +1,478 @@
+ * Currently implemented only for jemalloc. Used for online defragmentation. */
+void *defrag_jemalloc_alloc(size_t size) {
+    void *ptr = je_mallocx(size, MALLOCX_TCACHE_NONE);
+    return ptr;
+}
+void defrag_jemalloc_free(void *ptr, size_t size) {
+    if (ptr == NULL) return;
+    je_sdallocx(ptr, size, MALLOCX_TCACHE_NONE);
+}
+
+/* -----------------------------------------------------------------------------
+ * Helper functions for jemalloc translation between size and index
+ * -------------------------------------------------------------------------- */
+#define LG_QUANTOM_8_FIRST_POW2 3
+#define SIZE_CLASS_GROUP_SZ 4
+
+#define LG_QUANTOM_OFFSET_3 ((64 >> LG_QUANTOM_8_FIRST_POW2) - 1)
+#define LG_QUANTOM_OFFSET_4 (64 >> 4)
+
+#define get_binind_normal(_sz, _offset, _last_sz_pow2)                                                                 \
+    ((SIZE_CLASS_GROUP_SZ - (((1 << (_last_sz_pow2)) - (_sz)) >> ((_last_sz_pow2) - LG_QUANTOM_8_FIRST_POW2))) +       \
+     (((_last_sz_pow2) - (LG_QUANTOM_8_FIRST_POW2 + 3)) - 1) * SIZE_CLASS_GROUP_SZ + (_offset))
+/* Get the bin index in bin array from the reg_size.
+ *
+ * these are reverse engineered mapping of reg_size -> binind. We need this information because the utilization query
+ * returns the size of the buffer and not the bin index, and we need the bin index to access it's usage information
+ *
+ * Note: In case future PR will return the binind (that is better API anyway) we can get rid of
+ * these conversion functions*/
+inline unsigned jemalloc_sz2binind_lgq3(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 24, 32, 40, 48, 56, 64
+        return (sz >> 3) - 1;
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return get_binind_normal(sz, LG_QUANTOM_OFFSET_3, last_sz_in_group_pow2);
+}
+
+inline unsigned jemalloc_sz2binind_lgq4(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 32, 48, 64
+        return (sz >> 4);
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return get_binind_normal(sz, LG_QUANTOM_OFFSET_4, last_sz_in_group_pow2);
+}
+
+/* -----------------------------------------------------------------------------
+ * Get INFO string about the defrag.
+ * -------------------------------------------------------------------------- */
+/*
+ * add defrag info string into info
+ */
+sds defrag_jemalloc_get_fragmentation_info(sds info) {
+    if (!defrag_supported) return info;
+    je_binfo *binfo;
+    je_busage *busage;
+    unsigned nbins = arena_bin_conf.nbins;
+    if (nbins > 0) {
+        info = sdscatprintf(info,
+                            "jemalloc_quantom:%d\r\n"
+                            "hit_ratio:%lu%%,hits:%lu,misses:%lu\r\n"
```

Comment:
```suggestion
                            "defrag_hit_ratio:%lu%%\r\ndefrag_hits:%lu,defrag_misses:%lu\r\n"
```

This is a weird info syntax. We normally don't have multiple pairs on the same line separated by commands. I think each field should get its own line. Each info field needs to also stand on its own, some clients like python decompose each element into a key-value pair, and hit_ratio isn't a very descriptive name.

How is this from the other metrics like `ative_defrag_hits` ?

Reply:
I didn't want to change exiting defrag metrics tracked from `defrag.c` file. These (existing metrics) could be, potentially, for non jemalloc defrag and have upper level logic, like keys defragged. I can think of nice other metrics at the defrag.c level, like have information about types of keys that where defragged. Here this information is from the perspective of the allocator, this specific metric does correspond to the `active_defrag_hits` in existing code. @madolson Does this make sense? or do you still think we should consolidate?